File: backend/Services/Parsers/MurdleParser.cs
using System.Text.RegularExpressions;
using backend.Models;

namespace backend.Services.Parsers
{
    public class MurdleParser : BasicTimeResultParser
    {
        private ILogger<MurdleParser> _logger;

        public MurdleParser(ILogger<MurdleParser> logger) : base(logger)
        {
            _logger = logger;
        }

        public override bool CountWinner => true;
        public override string GameName => "Murdle";
        public override bool GolfScoring => true;
        public override string? HelpText => null;
        private const string ScoreGroup = "score";
        // 0\uFE0F\u20E3 = 0️⃣
        // 1\uFE0F\u20E3 = 1️⃣
        // 2\uFE0F\u20E3 = 2️⃣
        // 3\uFE0F\u20E3 = 3️⃣
        // 4\uFE0F\u20E3 = 4️⃣
        // 5\uFE0F\u20E3 = 5️⃣
        // 6\uFE0F\u20E3 = 6️⃣
        // 7\uFE0F\u20E3 = 7️⃣
        // 8\uFE0F\u20E3 = 8️⃣
        // 9\uFE0F\u20E3 = 9️⃣
        protected override Regex Parser => new Regex($@"{GameName} for [\d/]+[^✅❌]+(?<{ScoreGroup}>[✅❌]+)\s+(?<{TimeGroup}>[:0123456789\uFE0F\u20E3]+)", RegexOptions.Singleline);
        public override string Url => "https://murdle.com/";

        protected override string GetCleanResult(string result, Match parserResults)
        {
            var streakEmojisStart = result.IndexOf("⚖️");
            return result[..streakEmojisStart].Trim();
        }

        public override string? GetScoreValue(DailyResult dailyResult)
        {
            return dailyResult.Score?.ToString() + " - " + base.GetScoreValue(dailyResult);
        }

        protected override DailyResult SetScore(DailyResult dailyResult, Match parserResults)
        {
            if (parserResults.Groups.ContainsKey(ScoreGroup))
            {
                // main murdle score is the number of failures - fewer failures = better score
                dailyResult.Score = parserResults.Groups[ScoreGroup].Value.Split('❌').Length - 1;
            }

            if (parserResults.Groups.ContainsKey(TimeGroup))
            {
                // secondary murdle score is the time it took - less time elapsed = better score
                var timeSpan = parserResults.Groups[TimeGroup].Value
                    .Replace("\uFE0F", string.Empty)
                    .Replace("\u20E3", string.Empty);

                var time = ParseTimeSpanString(timeSpan, parserResults);
                if (time != null)
                {
                    dailyResult.Time = time;
                }
            }

            // overall score is the fewest failures with the best time
            return dailyResult;
        }
    }
}

File: lambdas/src/parsers/murdle-parser.ts
import { DailyResult } from '../models/daily-result';
import { ResultParser, getGroup, hasGroup } from './base/result-parser';

export class MurdleParser extends ResultParser {
  readonly countWinner = true;
  readonly gameName = 'Murdle';
  readonly golfScoring = true;
  readonly helpText = null;
  readonly url = 'https://murdle.com/';
  
  private static readonly SCORE_GROUP = 'score';
  private static readonly TIME_GROUP = 'time';
  
  // Matches "Murdle for 12/4/2025\n\n👤🔪🏡     🕰️\n✅✅✅     3️⃣:4️⃣3️⃣"
  protected readonly parser = new RegExp(
    `${this.gameName} for [\\d/]+[^✅❌]+(?<${MurdleParser.SCORE_GROUP}>[✅❌]+)\\s+(?<${MurdleParser.TIME_GROUP}>[:0123456789\uFE0F\u20E3]+)`,
    's'
  );
  
  protected getCleanResult(result: string, match: RegExpMatchArray): string {
    // Note: C# uses result[..streakEmojisStart] without checking if IndexOf returns -1
    // TS adds safety check to handle case where ⚖️ is not present
    const streakEmojisStart = result.indexOf('⚖️');
    if (streakEmojisStart > -1) {
      return result.substring(0, streakEmojisStart).trim();
    }
    return result.trim();
  }
  
  getScoreValue(dailyResult: DailyResult): string | null {
    const scoreStr = dailyResult.score?.toString() ?? '';
    const timeStr = dailyResult.time ?? '';
    return `${scoreStr} - ${timeStr}`;
  }
  
  protected setScore(dailyResult: DailyResult, match: RegExpMatchArray): DailyResult {
    if (hasGroup(match, MurdleParser.SCORE_GROUP)) {
      // Score is the number of failures (❌) - fewer failures = better
      const scoreStr = getGroup(match, MurdleParser.SCORE_GROUP) || '';
      dailyResult.score = scoreStr.split('❌').length - 1;
    }
    
    if (hasGroup(match, MurdleParser.TIME_GROUP)) {
      // Parse time with emoji digits
      const timeStr = (getGroup(match, MurdleParser.TIME_GROUP) || '')
        .replace(/\uFE0F/g, '')
        .replace(/\u20E3/g, '');
      
      const time = this.parseTimeSpanString(timeStr);
      if (time) {
        dailyResult.time = time;
      }
    }
    
    return dailyResult;
  }
  
  /**
   * Parses a time string into HH:MM:SS format
   * Supports: H:MM:SS, M:SS, and SS formats (matching C# BasicTimeResultParser)
   */
  private parseTimeSpanString(timeSpan: string): string | null {
    // Replace dots with colons (for alternate formats like 1.30)
    timeSpan = timeSpan.replace(/\./g, ':');
    
    const colonCount = (timeSpan.match(/:/g) || []).length;
    
    let hours = 0;
    let minutes = 0;
    let seconds = 0;
    
    if (colonCount === 2) {
      // H:MM:SS
      const parts = timeSpan.split(':');
      hours = parseInt(parts[0], 10);
      minutes = parseInt(parts[1], 10);
      seconds = parseInt(parts[2], 10);
    } else if (colonCount === 1) {
      // M:SS
      const parts = timeSpan.split(':');
      minutes = parseInt(parts[0], 10);
      seconds = parseInt(parts[1], 10);
    } else {
      // SS
      seconds = parseInt(timeSpan, 10);
    }
    
    if (isNaN(hours) || isNaN(minutes) || isNaN(seconds)) {
      return null;
    }
    
    return `${String(hours).padStart(2, '0')}:${String(minutes).padStart(2, '0')}:${String(seconds).padStart(2, '0')}`;
  }
}
